### ISSUE TITLE
Data distribution not well balanced in a cluster

### ISSUE THREAD

**Issue by fgelcer** [scylladb/scylla#6193]
*Installation details*
Scylla version (or git commit hash): 4.0.rc1-0.20200412.a8ef820f27c with build-id 15fa5c2e803b98279bb1932abf806dfb7937238c
Cluster size: 4 DB nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0e1584bf517753731

no nemesis run during the prepare data.

the seed node has much less data than other nodes.

nodetool status before preparing the data set:
```
=== stdout ===
Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.0.212.58   294.57 KB  256          ?       c5dc7721-d7c3-4887-aa2d-878e4528e1a2  1a
UN  10.0.218.141  298.75 KB  256          ?       9969cfb5-9db0-4d41-8aef-7da854b33a51  1a
UN  10.0.154.78   304.88 KB  256          ?       bd4e2772-aaea-4fdc-b239-dbbb349ab1ee  1a
UN  10.0.126.158  324.91 KB  256          ?       54b5f369-8a8e-4efe-aa65-4cdbcff16aca  1a

Note: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless
```

nodetool status after the data set was fully prepared:
```
=== stdout ===
Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.0.212.58   1001.4 GB  256          ?       c5dc7721-d7c3-4887-aa2d-878e4528e1a2  1a
UN  10.0.218.141  1 TB       256          ?       9969cfb5-9db0-4d41-8aef-7da854b33a51  1a
UN  10.0.154.78   1.02 TB    256          ?       bd4e2772-aaea-4fdc-b239-dbbb349ab1ee  1a
UN  10.0.126.158  428.78 GB  256          ?       54b5f369-8a8e-4efe-aa65-4cdbcff16aca  1a


Note: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless

```

[here](https://snapshot.raintank.io/dashboard/snapshot/DCduhfeBeyysyDHwLKtFpy2ogNfpK2NV) the grafana chart showing "disk size per instance"

[here](https://snapshot.raintank.io/dashboard/snapshot/UlV5fBbmEdJG94UMsSg91RRgJ1OrG74X?orgId=2) other charts of grafana

**Comment by slivne** [comment]
which longevity are you running large partition ?

**Comment by fgelcer** [comment]
> which longevity are you running large partition ?

1TB

**Comment by roydahan** [comment]
No, it's the regular 1TB longevity

**Comment by fgelcer** [comment]
c-s results of the prepare data set:
```
Results:
Op rate                   :   10,960 op/s  [WRITE: 10,960 op/s]
Partition rate            :   10,960 pk/s  [WRITE: 10,960 pk/s]
Row rate                  :   10,960 row/s [WRITE: 10,960 row/s]
Latency mean              :    4.6 ms [WRITE: 4.6 ms]
Latency median            :    4.2 ms [WRITE: 4.2 ms]
Latency 95th percentile   :    8.8 ms [WRITE: 8.8 ms]
Latency 99th percentile   :   20.9 ms [WRITE: 20.9 ms]
Latency 99.9th percentile :   40.0 ms [WRITE: 40.0 ms]
Latency max               : 1846.5 ms [WRITE: 1,846.5 ms]
Total partitions          : 50,000,000 [WRITE: 50,000,000]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 01:16:02

END
```
```
Results:
Op rate                   :    8,815 op/s  [WRITE: 8,815 op/s]
Partition rate            :    8,815 pk/s  [WRITE: 8,815 pk/s]
Row rate                  :    8,815 row/s [WRITE: 8,815 row/s]
Latency mean              :    5.7 ms [WRITE: 5.7 ms]
Latency median            :    4.3 ms [WRITE: 4.3 ms]
Latency 95th percentile   :   14.9 ms [WRITE: 14.9 ms]
Latency 99th percentile   :   31.1 ms [WRITE: 31.1 ms]
Latency 99.9th percentile :   53.0 ms [WRITE: 53.0 ms]
Latency max               : 2013.3 ms [WRITE: 2,013.3 ms]
Total partitions          : 50,000,000 [WRITE: 50,000,000]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 01:34:31

END
```
```
Results:
Op rate                   :    7,564 op/s  [WRITE: 7,564 op/s]
Partition rate            :    7,564 pk/s  [WRITE: 7,564 pk/s]
Row rate                  :    7,564 row/s [WRITE: 7,564 row/s]
Latency mean              :    6.6 ms [WRITE: 6.6 ms]
Latency median            :    4.5 ms [WRITE: 4.5 ms]
Latency 95th percentile   :   18.6 ms [WRITE: 18.6 ms]
Latency 99th percentile   :   36.2 ms [WRITE: 36.2 ms]
Latency 99.9th percentile :   69.2 ms [WRITE: 69.2 ms]
Latency max               : 2285.9 ms [WRITE: 2,285.9 ms]
Total partitions          : 50,000,000 [WRITE: 50,000,000]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 01:50:09

END
```
```
Results:
Op rate                   :    6,034 op/s  [WRITE: 6,034 op/s]
Partition rate            :    6,034 pk/s  [WRITE: 6,034 pk/s]
Row rate                  :    6,034 row/s [WRITE: 6,034 row/s]
Latency mean              :    8.3 ms [WRITE: 8.3 ms]
Latency median            :    4.9 ms [WRITE: 4.9 ms]
Latency 95th percentile   :   24.2 ms [WRITE: 24.2 ms]
Latency 99th percentile   :   43.5 ms [WRITE: 43.5 ms]
Latency 99.9th percentile :   87.1 ms [WRITE: 87.1 ms]
Latency max               : 2334.1 ms [WRITE: 2,334.1 ms]
Total partitions          : 50,000,000 [WRITE: 50,000,000]
Total errors              :          0 [WRITE: 0]
Total GC count            : 0
Total GC memory           : 0.000 KiB
Total GC time             :    0.0 seconds
Avg GC time               :    NaN ms
StdDev GC time            :    0.0 ms
Total operation time      : 02:18:06

END
```

**Comment by fgelcer** [comment]
this is the reproducer job
https://jenkins.scylladb.com/view/scylla-4.0/job/scylla-4.0/job/reproducers/job/longevity-1tb-7days

**Comment by slivne** [comment]
what is the schema ?

![disk_size](https://user-images.githubusercontent.com/3465480/79159447-c1746b00-7de0-11ea-9e18-f142bf418cf1.png)

~20:15 you have a node that was removed ?

~20:30 you have another node being added ?

can you clarify this please

**Comment by fgelcer** [comment]
yes, after the prepare was done, there was a nemesis (that is still running at this very moment) that terminate and replace node

**Comment by slivne** [comment]
can you please validate that all have auto_bootstrap=true set correctly

**Comment by fgelcer** [comment]
> can you please validate that all have auto_bootstrap=true set correctly

nope, the nodes for some reason are set as ```auto_bootstrap: false```

**Comment by fgelcer** [comment]
nodetool describering keyspace1:

[nodetool_describering.log](https://github.com/scylladb/scylla/files/4471908/nodetool_describering.log)

**Comment by dorlaor** [comment]
Did you also run nodetool cleanup?

**Comment by slivne** [comment]
indeed something is off here:

```
seed node 10.0.126.158

du -h /var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/pending_delete
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/upload
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/staging
415G	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/

nodetool cfstats keyspace1
Total number of tables: 54
----------------
Keyspace : keyspace1
	Read Count: 17495492
	Read Latency: 4.8042078496563574E-4 ms
	Write Count: 415355328
	Write Latency: 1.2233157148762998E-5 ms
	Pending Flushes: 0
		Table: standard1
		SSTable count: 2671
		SSTables in each level: [12/4, 143/10, 0, 2516/1000]
		Space used (live): 445039965452
		Space used (total): 445039965452
		Space used by snapshots (total): 0
		Off heap memory used (total): 3398163904
		SSTable Compression Ratio: 0.0
		Number of partitions (estimate): 408753751
		Memtable cell count: 1736297
		Memtable data size: 2746819368
		Memtable off heap memory used: 3391225856
		Memtable switch count: 1172
		Local read count: 17495495
		Local read latency: 0.480 ms
		Local write count: 415355341
		Local write latency: 0.012 ms
		Pending flushes: 0
		Percent repaired: 0.0
		Bloom filter false positives: 32512
		Bloom filter false ratio: 0.01316
		Bloom filter space used: 2544232
		Bloom filter off heap memory used: 2544176
		Index summary off heap memory used: 4393872
		Compression metadata off heap memory used: 0
		Compacted partition minimum bytes: 925
		Compacted partition maximum bytes: 1109
		Compacted partition mean bytes: 1109
		Average live cells per slice (last five minutes): 0.0
		Maximum live cells per slice (last five minutes): 0
		Average tombstones per slice (last five minutes): 0.0
		Maximum tombstones per slice (last five minutes): 0
		Dropped Mutations: 0

----------------


```

```
other node 10.0.218.141
 du -h /var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/pending_delete
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/0000000000050613.sstable
i0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/0000000000057000.sstable
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/upload
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/staging
0	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/0000000000102404.sstable
f997G	/var/lib/scylla/data/keyspace1/standard1-455a63707d8711ea93c5000000000009/

nodetool cfstats keyspace1 
Total number of tables: 54
----------------
Keyspace : keyspace1
	Read Count: 44757162
	Read Latency: 0.002252133256349006 ms
	Write Count: 1019524827
	Write Latency: 1.1516422836459283E-5 ms
	Pending Flushes: 0
		Table: standard1
		SSTable count: 6305
		SSTables in each level: [13/4, 140/10, 0, 6152/1000]
		Space used (live): 1069415190452
		Space used (total): 1069415190452
		Space used by snapshots (total): 0
		Off heap memory used (total): 1924210880
		SSTable Compression Ratio: 0.0
		Number of partitions (estimate): 982249524
		Memtable cell count: 1124029
		Memtable data size: 1778291396
		Memtable off heap memory used: 1917190144
		Memtable switch count: 2422
		Local read count: 44759071
		Local read latency: 2.252 ms
		Local write count: 1019526219
		Local write latency: 0.012 ms
		Pending flushes: 0
		Percent repaired: 0.0
		Bloom filter false positives: 20005
		Bloom filter false ratio: 0.01074
		Bloom filter space used: 2626920
		Bloom filter off heap memory used: 2626864
		Index summary off heap memory used: 4393872
		Compression metadata off heap memory used: 0
		Compacted partition minimum bytes: 925
		Compacted partition maximum bytes: 1109
		Compacted partition mean bytes: 1109
		Average live cells per slice (last five minutes): 0.0
		Maximum live cells per slice (last five minutes): 0
		Average tombstones per slice (last five minutes): 0.0
		Maximum tombstones per slice (last five minutes): 0
		Dropped Mutations: 0

----------------

```

so on the seed there is ~50% of the files - and they do not fit into memory - something is broken

**Comment by slivne** [comment]
![writes_per_instance_replica](https://user-images.githubusercontent.com/3465480/79165891-b9222d00-7dec-11ea-848d-40ecae6ca2ed.png)

the green line is the seed node - its receiving much less replica writes - which aligns with the fact it has much less data - but doesn't explain why this happens.


```
[root@longevity-tls-1tb-7d-longevit-db-node-a9c535be-2 scylla]# nodetool describering keyspace1 | grep 10.0.126.158 | wc -l
755
[root@longevity-tls-1tb-7d-longevit-db-node-a9c535be-2 scylla]# nodetool describering keyspace1 | grep 10.0.218.141 | wc -l
787
[root@longevity-tls-1tb-7d-longevit-db-node-a9c535be-2 scylla]# nodetool describering keyspace1 | grep 10.0.154.78 | wc -l
788
[root@longevity-tls-1tb-7d-longevit-db-node-a9c535be-2 scylla]# nodetool describering keyspace1 | grep 10.0.212.58 | wc -l
742
```

checking 

```
cat /etc/scylla/scylla.yaml | grep exp
experimental: true
```

could this be related to the token selection algorithms ?

**Comment by slivne** [comment]
could it be related to this https://github.com/scylladb/scylla/pull/5485/commits/32f4489a1879ed29fa82315e45c5adbe08f09a44#

@roydahan / @fgelcer if this reproduces for you - e.g the seed is allways with less data.

Please disable experimental flag (LWT is not experimental anymore) - and lets see if this fixes the problem.

I recall vaguely that CDC had to change the way that tokens had to be selected.

**Comment by slivne** [comment]
@kbr- / @haaawk does the patch 32f4489 change the way tokens are assigned to nodes ?

**Comment by haaawk** [comment]
> @kbr- / @haaawk does the patch [32f4489](https://github.com/scylladb/scylla/commit/32f4489a1879ed29fa82315e45c5adbe08f09a44) change the way tokens are assigned to nodes ?

As far as I know, no. Tokens should be assigned in exactly the same way. @kbr- will know better.

**Comment by slivne** [comment]
There are two issues:

- We need to figure out why we have incorrect distribution - we will try to see if we can bisect this with dtest. 

- We need to understand nodetool describring - the nmber of replicas each node should exist in is 768 in all cases.

**Comment by slivne** [comment]
QA will try to see if they can help find the source of the issues

**Comment by enaydanov** [comment]
I've run modified SCT 100gb longevity test with 4 DB nodes, RF=3, LCS compaction strategy, NoOpMonkey nemesis and got following results:

AMI mentioned in this issue (ami-0e1584bf517753731): uneven distribution
```
$ nodetool status

Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.0.171.50   101.83 GB  256          ?       d63c638f-7888-482a-9e11-bed154935531  1a
UN  10.0.206.100  35.11 GB   256          ?       daf3394d-5aa5-4f46-9df7-13b23dbdf223  1a
UN  10.0.204.7    94.62 GB   256          ?       f66e2e63-dd2f-4005-8146-4770f9c91afd  1a
UN  10.0.120.89   93.85 GB   256          ?       7bd82e7e-7d54-4fc6-b45d-748825bc82e2  1a
```

3.3.0 AMI (ami-00eb480a240d1955e): even distribution
```
$ nodetool status

Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.0.87.32    83.03 GB   256          ?       8443f231-206d-4404-a089-432a33e68bc3  1a
UN  10.0.140.48   74.55 GB   256          ?       7b6bf661-40dc-4d43-9941-d459371d3554  1a
UN  10.0.149.19   82.11 GB   256          ?       6228208b-c2c4-46ba-af8f-beada76c68dc  1a
UN  10.0.174.139  76.89 GB   256          ?       49b46d9e-09cf-45cd-a290-a71479b48597  1a
```

4.0.rc0 AMI (ami-02a85d5a79509d2d5): run in progress

**Comment by enaydanov** [comment]
Issue reproduced on 4.0.rc0 (ami-02a85d5a79509d2d5):

```
$ nodetool status

Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
UN  10.0.15.114  36.53 GB   256          ?       98429fc3-1e89-4029-ac1c-325179752142  1a
UN  10.0.126.57  88.17 GB   256          ?       aea7e0f2-c2c3-4dc6-8ffd-8eda27f4ab8e  1a
UN  10.0.74.155  90.62 GB   256          ?       f2df2267-b8d1-4a1b-a5d8-a6c57a289f44  1a
UN  10.0.65.254  101.39 GB  256          ?       20eca592-3eda-478b-b9c8-03266879b8ba  1a
```

**Comment by slivne** [comment]
Thanks @enaydanov - so we know its a regression now the hard task of finding the bad commit on master :(

**Comment by bentsi** [comment]
on 4.0.rc0 (ami-02a85d5a79509d2d5) distribution is also uneven (copying from Evgeniy's email) 
```
$ nodetool status

Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address      Load       Tokens       Owns    Host ID                               Rack
UN  10.0.15.114  36.53 GB   256          ?       98429fc3-1e89-4029-ac1c-325179752142  1a
UN  10.0.126.57  88.17 GB   256          ?       aea7e0f2-c2c3-4dc6-8ffd-8eda27f4ab8e  1a
UN  10.0.74.155  90.62 GB   256          ?       f2df2267-b8d1-4a1b-a5d8-a6c57a289f44  1a
UN  10.0.65.254  101.39 GB  256          ?       20eca592-3eda-478b-b9c8-03266879b8ba  1a
```

**Comment by fgelcer** [comment]
Is it on seed node as we seen on 1tb job?

**Comment by aleksbykov** [comment]
Issue reproduced with dtest on branch-4.0 (unstable/branch-4.0:2020-04-14T10%3A17%3A01Z):
```
'Status=Up/Down\n'
  '|/ State=Normal/Leaving/Joining/Moving\n'
  '--  Address    Load       Tokens       Owns    Host ID                               Rack\n'
  'UN  127.0.0.1  173.24 MB  256          ?       67d41c96-ce44-4bb0-8d45-cb881e983cce  rack1\n'
  'UN  127.0.0.2  43.9 MB    256          ?       0737f018-59b4-408d-815f-1e190e2775d5  rack1\n'
  'UN  127.0.0.3  173.23 MB  256          ?       41ae35f6-ef44-4d9e-9dfb-9e7d4c55893f  rack1\n'
  'UN  127.0.0.4  171.58 MB  256          ?       b8cf21fe-42f1-4159-b513-4a8c61e66f1f  rack1\n'
  ```

with 3.3 is not reproduced (unstable/branch-3.3:2020-04-13T20%3A29%3A26Z):
```
Datacenter: datacenter1
=======================
Status=Up/Down
| State=Normal/Leaving/Joining/Moving
--  Address    Load       Tokens       Owns    Host ID                               Rack
UN  127.0.0.1  734.56 MB  256          ?       0e196b6f-513c-4ac7-8fa2-19f71a3002d2  rack1
UN  127.0.0.2  806.82 MB  256          ?       2fd2e3f9-c9b8-42a0-8e48-866bf6851768  rack1
UN  127.0.0.3  752.3 MB   256          ?       71d279c1-16e3-4c94-b318-a4e1ed5eedff  rack1
UN  127.0.0.4  860.37 MB  256          ?       4ec442e4-63f8-4abb-8b36-6f04d3c01a0e  rack1
```

**Comment by slivne** [comment]
Lets start with the point we switched the AMI images, lets find the last master ami prior to switching to macine-image and lets see if the issue reproduces on that one.

**Comment by kbr-** [comment]
> @kbr- / @haaawk does the patch [32f4489](https://github.com/scylladb/scylla/commit/32f4489a1879ed29fa82315e45c5adbe08f09a44) change the way tokens are assigned to nodes ?

No, the CDC patches didn't change the way tokens are assigned. At least not intentionally.

**Comment by aleksbykov** [comment]
Issue already reproduced on this build:
```
scylla-product: scylla
scylla-release: 0.20200325.78f5afec30
scylla-version: 4.0.rc0
scylla-BuildID[sha1]: 6a210b1adac28793113c61e037cdd06040f0d217
```

**Comment by slivne** [comment]
@roydahan / @aleksbykov please prioritize this - I need our help to find the commit (as a starting point ranges) that caused this regression - please use the relocatable repo and try and bisect. 
between (3.3.0 - 4.0.rc0)
- 3.3.0 - doesn't have the issue, 
- 4.0.rc0 - has the issue.

Thanks,

Shlomi

**Comment by aleksbykov** [comment]
Issue was found on next relocatable package on master 
scylla-product: scylla
scylla-release: 0.20200206.27b36beb4ab
scylla-version: 666.development
scylla-BuildID[sha1]: 3b5f026ff5948080990403bca77313d71dbea59d
scylla-debug-BuildID[sha1]: e486d8a7a3399e22f0d360f274f9047dd272bd48

**Comment by gcarmin** [comment]
I just hit this issue with a benchmark testing.
4Xi3.metal on AMI 4.0rc1 ScyllaDB 4.0.rc1 (ami-0ddd86daaba39d1d1)
8Xm5.8xlarge Centos7 with cassandra-stress from scylla 4.0rc1.
All on the same AZ
I run the following command:
`nohup cassandra-stress write no-warmup n=4856250000 cl=one -mode native cql3 -node $IPS -rate threads=200 -log file=[log_file] -pop seq=1..4856250000 > loading.data.log &`
I got one node with more data as you scan see in the attached screenshots:
![Screenshot_2020-04-16 Overview - Grafana](https://user-images.githubusercontent.com/13367089/79445741-d1a96780-7fe5-11ea-909c-31548b8024a4.png)
![Screenshot_2020-04-16 OS Metrics - Grafana](https://user-images.githubusercontent.com/13367089/79445745-d2da9480-7fe5-11ea-9bc7-4abf03e25726.png)
![Screenshot_2020-04-16 I O - Grafana](https://user-images.githubusercontent.com/13367089/79445747-d40bc180-7fe5-11ea-8ea8-55e730f7eeea.png)
![Screenshot_2020-04-16 Detailed - Grafana](https://user-images.githubusercontent.com/13367089/79445748-d4a45800-7fe5-11ea-94be-531694fdb741.png)
![Screenshot_2020-04-16 CPU Metrics - Grafana](https://user-images.githubusercontent.com/13367089/79445750-d53cee80-7fe5-11ea-89d7-99ca6cb1ebc7.png)
![image](https://user-images.githubusercontent.com/13367089/79445753-d5d58500-7fe5-11ea-9b74-565fcfa8ad80.png)

**Comment by slivne** [comment]
@alexbykov - thanks

@bhalevy - we need someone from your team to help find the bug so we can then get someone to fix it.

please look at https://github.com/scylladb/scylla/issues/6193#issuecomment-614410792 for the info of the first bad commit.

**Comment by bhalevy** [comment]
what dtest reproduces this?  Can you please provide a command line?

**Comment by aleksbykov** [comment]
@bhalevy , you can find command and dtest for repo here 
https://github.com/scylladb/scylla-dtest/pull/1428

**Comment by bhalevy** [comment]
I bisected using @aleksbykov's dtest and the regression starts at 31593e1451 Merge "Change token representation to int64_t" from Piotr

**Comment by bhalevy** [comment]
@haaawk can you please look into it?

**Comment by haaawk** [comment]
Hmm. This is really surprising. I will have a look.

**Comment by kbr-** [comment]
Interesting!

It's curious that we don't have any dtests that would immediately catch such an issue. We should think of something so that this regression doesn't happen again.

Idea for a test: send a bunch of writes to random partitions, and check whether they were saved to the replicas that we expect them to be saved, according to murmur3-calculated tokens of these writes (calculated by the test, not by Scylla).

**Comment by slivne** [comment]
now we do have a dtest - thats how it was bisected

**Comment by kbr-** [comment]
> now we do have a dtest - thats how it was bisected

Right. Somehow my brain completely ignored that fact...

**Comment by espindola** [comment]
9c202b52da79aeece9814ebfc922951ad51f7317 is the first commit in the merge that fails.

**Comment by espindola** [comment]
I should have a patch in a minute.

**Comment by espindola** [comment]
patch sent.

**Comment by haaawk** [comment]
Wow. What's going on here? It seems that
```
int64_t token_value = (int64_t)dht::get_random_number<uint64_t>()
```
 is ok but
```
int64_t token_value = dht::get_random_number<int64_t>()
```
 is not. get_random_number is defined as:
```
template <typename T>
inline auto get_random_number() {
    static thread_local std::default_random_engine re{std::random_device{}()};
    static thread_local std::uniform_int_distribution<T> dist{};
    return dist(re);
}
```
Why wouldn't it work with int64_t?

I found the answer on [cppreference](https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution):

> ### Template parameters
> **IntType** - The result type generated by the generator. The effect is undefined if this is not one of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long.

signed long long is not one of the types that guarantee defined behavior :(.

Sorry for introducing this bug. I wish there was a static_assert checking the template parameter type inside uniform_int_distribution.

**Comment by slivne** [comment]
Patch on ML "[PATCH] dht: Use get_random_number<uint64_t> instead of int64_t in token::get_random_token"

**Comment by avikivity** [comment]
Wow, well done.

**Comment by avikivity** [comment]
@haaawk the problem is not `signed long long`, it's the constructor. get_random_number uses the default constructor, which constructs from 0 to numeric_limits::max(). So we lost all the negative numbers.

Note that get_random_token() has a bug - it can generate std::numeric_limits<int64_t>::min(), which is not a legal token. Please fix.

**Comment by roydahan** [comment]
@kbr- i was also really surprised we don’t have any dtest to catch something like that.
(I expected something to exist from the early days).

We will add some more dtest around these, but it would be good if you can send your idea as well.

**Comment by avikivity** [comment]
Backported to 4.0, 3.3, 3.2.

**Comment by avikivity** [comment]
Backported to 4.0 (only vulnerable branch)

**Comment by espindola** [comment]
> @haaawk the problem is not `signed long long`, it's the constructor. get_random_number uses the default constructor, which constructs from 0 to numeric_limits::max(). So we lost all the negative numbers.
> 
> Note that get_random_token() has a bug - it can generate std::numeric_limits<int64_t>::min(), which is not a legal token. Please fix.

We are now calling normalize which maps min to max. Since we are using uniform_int_distribution int distribution it would probably be better to exclude min directly, no?

**Comment by fgelcer** [comment]
it looks like it is fixed on 4.0.rc2-0.20200421.89e79023aeb:
```
=== stdout ===
Datacenter: eu-west
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address       Load       Tokens       Owns    Host ID                               Rack
UN  10.0.239.112  1.16 GB    256          ?       e86745c1-1a0c-4a70-abdc-cd90412c1b3c  1a
UN  10.0.22.81    1.18 GB    256          ?       2b49418b-0926-4b0d-a33c-ee1bff6e8adc  1a
UN  10.0.70.195   1.16 GB    256          ?       6c58aace-8380-4090-a39a-031feb6b2931  1a
UN  10.0.136.134  1.4 GB     256          ?       de5c6362-bd96-499f-92f5-60888796b687  1a
UN  10.0.235.88   1.19 GB    256          ?       a7b0d8bf-d752-41ca-a96e-b2a494e16da7  1a
UN  10.0.18.10    1.58 GB    256          ?       bace33fb-9b71-4677-880f-9f82e3316719  1a

Note: Non-system keyspaces don't have the same replication settings, effective ownership information is meaningless
```

**Comment by gcarmin** [comment]
when are we going to have AWS AMI 4.0 rc2?
